design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+* self-terminating or runs "forever" (i.e., until terminated by a user)
+* query input types; note, we extend the discussion to arbitrary n-ary queries, even if KSQL only supports unary and binary queries atm
+  (why we consider only those two categories gets clear later):
+  * all inputs are tables
+  * at least one input is a stream
+* we don't consider the output type, because it's not a property of the query itself, but derived by the design of query language
+* we don't consider transient (client/CLI) queries vs. persistent queries (CSAS, CTAS) because both should have the same semantics
+
+Note, that while we consider those properties as independent in general, there are some corner case for which some properties contradict each other,
+i.e., there are some partial dependencies. More details below.
+
+Before we dive into a detailed discussion, we lay out (in less technical terms) what type of queries we want to support (this is for sure an incomplete list):
+
+* forever running query over an input STREAM that returns an output stream
+  * think: simple stream filter, stream-table join, or stream-stream join
+  * this may be a transient client query or a persistent query
+* forever running aggregation query over an input STREAM that returns an output table
+  * think: a persistent continuous query (may be windowed)
+  * for a persistent query, this should populate a MV
+  * for a transient client query, we would like to get the corresponding changelog stream
+* a RDBMS style query against a TABLE that terminates automatically and returns a table (called 'pull queries' in KLIP-8)
+  * conceptually, we can think of everything that 'classic' SQL allows
+* a self-terminating query over an input STREAM that returns an output stream
+  * think of a query with a range in time, e.g., yesterday, or everything up to "now" (but don't wait for future events)
+* a forever running query over an input TABLE that return a changelog stream
+  * answer my query on the current table state, and keep updating the answer if the base table state changes
+
+### Baseline Assumptions
+
+* Data stored in Kafka topics is just data; there is no semantic interpretation attached
+* A _STREAM_ is an unbounded append only-sequence of immutable facts
+* A _TABLE_ is a mutable bounded set
+  * for each point in (event) time, there exist exactly one TABLE _version_ (i.e., a snapshot)
+  * a TABLE _evolves_ over (event) time (we call it an _evolving table_) when new insert/updates/deletes are applied
+    * a TABLE version is "complete" if all updates for the corresponding event time are applied
+    * updating a TABLE from version v to v+1 might imply multiple updates and might take some time,
+      hence, at any point in wall-clock time a TABLE might be in an inconsistent state
+      (we call those the _generation_ of a table, i.e., for each table version, there might be multiple generation and only the last generation is consistent).
```

Comment:
Not sure I full follow what these generations are. Can you add an example / elaborate?

Reply:
Let's say you have the following input changelog topic that you want to materialize as table state:
```
<key,value,timestamp>
<A,1,100> <B,2,100>, <A,5,200>, <B,6,200>
```

There are two table _versions_:
```
t=100    t=200
A = 1    A = 5
B = 2    B = 6
```
But each _versions_ has two _genrations_ (ie, each applied input record, advance the generation)
```
t=100    t=100     t=200     t=200
g=0      g=1       g=2       g=3
A = 1    A = 1     A = 5     A = 5
         B = 2     B = 2     B = 6
```
g=0 represent an _incomplete_ state of version=100 because not all updates for version =100 are applied yet. Similarly, g=2 represents an incomplete state of version=200 for the same reason.

Using the term "version" vs "generation" makes it simple to talk about semantics, because it highlights the difference between processing time and event-time. For example, we can say that a pull query is always executed at the latest version and generation when the query is submitted. Or if we introduce `AS OF <timestamp>` keyword, we would give a guarantee about the table _version_ that is queried (but not necessarily which generation within the version is queried etc). We can also define "current version" as the materialized version when a query is executed (in system time) vs "latest version" as whatever if available in the topic (assuming the table materialization lags).

Does this make sense?

Reply:
Yep, thanks.  IMHO this is worth including in the klip

Reply:
Ack. I'll extend the KLIP to add those details.

About the second point: it depends what guarantees/semantics we want to support. If we say (what I would propose atm) `AS OF <timestamp>` give you latest generation of the specified version (ie, timestamp), we don't really need it because we don't give any guarantee that the version is final.